localstack/http/asgi.py

```diff
@@ -123,111 +137,56 @@ def __init__(
-        if not self._more_body:
-            return 0, False
-
-        recv_future = asyncio.run_coroutine_threadsafe(self.receive(), self.event_loop)
-        event = recv_future.result()
-        # TODO: disconnect events
-        body = event["body"]
-        more = event.get("more_body", False)
-        buf.extend(body)
-        self._more_body = more
-        return len(body), more
-
-    def read(self, size: t.Optional[int] = None) -> bytes:
-        """
-        Reads up to ``size`` bytes from the object and returns them. As a convenience, if ``size`` is unspecified or
-        ``-1``, all bytes until EOF are returned. Like RawIOBase specifies, only one system call is ever made (in
-        this case, a call to the ASGI receive callable). Fewer than ``size`` bytes may be returned if the underlying
-        call returns fewer than ``size`` bytes.
-
-        :param size: the number of bytes to read
-        :return:
-        """
-        buf = self._buffer
+    def readable(self) -> bool:
+        return True
 
-        if not buf and not self._more_body:
-            return b""
-
-        if size is None or size == -1:
-            while True:
-                read, more = self._read_into(buf)
-                if not more:
-                    break
-
-            arr = bytes(buf)
-            buf.clear()
-            return arr
-
-        if len(buf) < size:
-            self._read_into(buf)
-
-        copy = bytes(buf[:size])
-        self._buffer = buf[size:]
-        return copy
+    def readinto(self, buf: bytearray | memoryview) -> int:
+        if not self._more_body:
+            return 0
 
-    def readline(self, size: t.Optional[int] = None) -> bytes:
-        buf = self._buffer
-        size = size if size is not None else -1
+        # max bytes we can write into the buffer
+        buf_size = len(buf)
 
-        while True:
-            i = buf.find(b"\n")  # FIXME: scans the whole buffer every time
+        # _buffered_body holds the carry-over of what we didn't read in the last iteration
+        if self._buffered_body is None:
+            # read from the underlying socket stream
+            recv_future = asyncio.run_coroutine_threadsafe(self.receive(), self.event_loop)
+            event = recv_future.result()
+            # TODO: disconnect events
+            more = event.get("more_body", False)
 
-            if i >= 0:
-                if 0 < size < i:
-                    break  # no '\n' in range
-                else:
-                    arr = bytes(buf[: (i + 1)])
-                    self._buffer = buf[(i + 1) :]
-                    return arr
-
-            # ensure the buffer has at least `size` bytes (or all)
-            if size > 0:
-                if len(buf) >= size:
-                    break
-            _, more = self._read_into(buf)
             if not more:
-                break
+                self._more_body = False
+                return 0
```

Comment:
question: should we return here? doesn't `more_body` indicates that there won't be more, but couldn't there be data in that last chunk? I think in the previous implementation, we would still extend the buffer with the body from this event.

Reply:
empirically i validated that there is always a distinct event with a `b''` body and `more_body=False`. in other words, there will _always_ be at least two calls to `receive()`: one to get the body (which will have `more_body=True`, even though there's nothing else to read), and then one with a zero body and `more_body=False`. :shrug:

Reply:
Oh okay, yeah I tried to read the ASGI documentation and they were not explicit about the fact that there would be a body in the event with `more_body=False`. Good to know!

Reply:
yep same. i'm not 100% confident it's in the spec, but at least hypercorn behaves in this way.

Reply:
perfect, well, that simplifies it and we don't need to check if there's a body then. Good for us 😄